### PR TITLE
Refactor ubuntu setup

### PR DESCRIPTION
**Pull request by cpuguy83**
Extracts all the jammy logic into reusable components.
See individual commits for more details.

### CODE REVIEW

File: frontend/deb/distro/install.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		const installScript = `#!/usr/bin/env sh
+set -ex
+
+# Make sure any cached data from local repos is purged since this should not
+# be shared between builds.
+rm -f /var/lib/apt/lists/_*
+apt autoclean -y
+
+apt update
+apt install -y $@
```
Comment (pmengelbert) [inline frontend/deb/distro/install.go:30]:
```suggestion
apt install -y "$@"
```

File: frontend/deb/distro/install.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+`
+		script := llb.Scratch().File(
+			llb.Mkfile("install.sh", 0o755, []byte(installScript)),
+			dalec.WithConstraint(&ei.Constraints),
+		)
+
+		p := "/tmp/dalec/internal/deb/install.sh"
+		llb.AddMount(p, script, llb.SourcePath("install.sh")).SetRunOption(ei)
```
Comment (pmengelbert) [inline frontend/deb/distro/install.go:38]: What does `llb.SourcePath` do?

File: frontend/deb/distro/install.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+// In some cases, with strict version constraints in the package's dependencies,
+// this will use `aptitude` to help resolve those dependencies since apt is
+// currently unable to handle strict constraints.
+//
+// This returns an [llb.RunOption] but it does create some things internally,
+// so any [llb.ConstraintsOpt] you want to pass in should come before this option
+// in a call to [llb.State.Run].
+func InstallLocalPkg(pkg llb.State) llb.RunOption {
+	return dalec.RunOptFunc(func(ei *llb.ExecInfo) {
+		// The apt solver always tries to select the latest package version even when constraints specify that an older version should be installed and that older version is available in a repo.
```
Comment (pmengelbert) [inline frontend/deb/distro/install.go:55]: nit (aesthetics): shorten these lines to ~90 characters or so

File: frontend/deb/distro/install.go (hunk 2 above)
Comment (cpuguy83) [reply frontend/deb/distro/install.go:38]: Let selects a subpath, and in this caes lets us mount just the file.

File: frontend/deb/distro/install.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	"github.com/Azure/dalec/frontend/deb"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/pkg/errors"
+)
+
+// AptInstall returns an [llb.RunOption] that uses apt to install the provided
+// packages.
+//
+// This returns an [llb.RunOption] but it does create some things internally,
+// so any [llb.ConstraintsOpt] you want to pass in should come before this option
```
Comment (pmengelbert) [inline frontend/deb/distro/install.go:17]: I really don't like the requirement that this has to follow all `llb.ConstraintsOpt`s. It introduces an implicit parameter structure for arguments to a function, that is not enforced in any way. If the caller makes a mistake, the result is a silent and subtle bug.

If we forget to tell a new developer about this, and they miss the comment, we could be in for a bad time (or at least a lost and frustrated developer)

File: frontend/azlinux/handler.go [hunk 5]
```diff
@@ -21,10 +21,9 @@ const (
 type installFunc func(context.Context, gwclient.Client, dalec.SourceOpts) (llb.RunOption, error)
 
 type worker interface {
-	Base(sOpt dalec.SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error)
+	Base(sOpt dalec.SourceOpts, opts ...llb.ConstraintsOpt) llb.State
```
Comment (pmengelbert) [inline frontend/azlinux/handler.go:24]: Why remove the error and silently return `llb.Scratch()`?

File: frontend/build.go [hunk 6]
```diff
@@ -94,3 +95,21 @@ func BuildWithPlatform(ctx context.Context, client gwclient.Client, f PlatformBu
...
 	return rb.Finalize()
 }
+
+// GetBaseImage returns an image that first checks if the client provided the
+// image in the build context matching the image ref.
+//
+// This follows the behavior of of the dockerfile frontend.
+func GetBaseImage(sOpt dalec.SourceOpts, ref string) llb.State {
+	return llb.Scratch().Async(func(ctx context.Context, _ llb.State, c *llb.Constraints) (llb.State, error) {
+		fromClient, err := sOpt.GetContext(ref, dalec.WithConstraint(c))
```
Comment (pmengelbert) [inline frontend/build.go:105]: Does this allow the user to pass in a base image as a build context?

**Review by pmengelbert — CHANGES_REQUESTED** [review]
Good work overall, I really think we need to find a solution for installing packages without introducing implicit parameter ordering.

File: frontend/build.go (hunk 6 above)
Comment (cpuguy83) [reply frontend/build.go:105]: Yes, this is something we already do.

File: frontend/azlinux/handler.go (hunk 5 above)
Comment (cpuguy83) [reply frontend/azlinux/handler.go:24]: Good find, I don't know what was going on here...

**Comment by cpuguy83** [comment]
All updated.

**Comment by pmengelbert** [comment]
Needs a rebase and it's failing a test, but I'll approve once that's sorted.

**Comment by cpuguy83** [comment]
This is rebased.
The failure was a network flake.

File: frontend/deb/distro/container.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+	debug := llb.Scratch().File(llb.Mkfile("debug", 0o644, []byte(`debug=2`)), opts...)
+	opts = append(opts, dalec.ProgressGroup("Install spec package"))
+
+	return baseImg.Run(
+		dalec.WithConstraints(opts...),
+		withRepos,
+		llb.AddEnv("DEBIAN_FRONTEND", "noninteractive"),
+		dalec.WithMountedAptCache(c.AptCachePrefix),
+		llb.AddMount("/etc/dpkg/dpkg.cfg.d/99-dalec-debug", debug, llb.SourcePath("debug"), llb.Readonly),
```
Comment (adamperlin) [inline frontend/deb/distro/container.go:41]: Can you add a comment explaining this file?

File: frontend/deb/distro/distro.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	return &img, nil
+}
+
+func (cfg *Config) RepoMounts(repos []dalec.PackageRepositoryConfig, sOpt dalec.SourceOpts, opts ...llb.ConstraintsOpt) (llb.RunOption, error) {
+	opts = append(opts, dalec.ProgressGroup("Prepare custom repos"))
+
+	repoConfig := cfg.RepoPlatformConfig
+	if repoConfig == nil {
+		repoConfig = &dalec.RepoPlatformConfig{
```
Comment (adamperlin) [inline frontend/deb/distro/distro.go:67]: Can you add this as a global potentially?

File: frontend/deb/distro/install.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+			Version:  spec.Version,
+			Revision: spec.Revision,
+			Dependencies: &dalec.PackageDependencies{
+				Runtime: buildDeps,
+			},
+			Description: "Build dependencies for " + spec.Name,
+		}
+
+		return in.Async(func(ctx context.Context, in llb.State, c *llb.Constraints) (llb.State, error) {
+			opts := append(opts, dalec.ProgressGroup("Insall build dependencies"))
```
Comment (adamperlin) [inline frontend/deb/distro/install.go:134]:
```suggestion
			opts := append(opts, dalec.ProgressGroup("Install build dependencies"))
```

File: frontend/deb/distro/install.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+set -ex
+
+# Make sure any cached data from local repos is purged since this should not
+# be shared between builds.
+rm -f /var/lib/apt/lists/_*
+apt autoclean -y
+
+apt update
+
+if ! command -v apititude > /dev/null; then
```
Comment (adamperlin) [inline frontend/deb/distro/install.go:78]:
```suggestion
if ! command -v aptitude > /dev/null; then
```

**Comment by cpuguy83** [comment]
Rebased and comments addressed.

File: frontend/deb/distro/install.go (hunk 1 above)
Comment (cpuguy83) [reply frontend/deb/distro/install.go]: TIL

**Comment by cpuguy83** [comment]
Rebased again.

**Review by jeremyrickard — APPROVED** [review]

File: frontend/deb/distro/container.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+		withRepos,
+		llb.AddEnv("DEBIAN_FRONTEND", "noninteractive"),
+		dalec.WithMountedAptCache(c.AptCachePrefix),
+		// This file makes dpkg give more verbose output which can be useful when things go awry.
+		llb.AddMount("/etc/dpkg/dpkg.cfg.d/99-dalec-debug", debug, llb.SourcePath("debug"), llb.Readonly),
+		dalec.RunOptFunc(func(cfg *llb.ExecInfo) {
+			tmp := llb.Scratch().File(llb.Mkfile("tmp", 0o644, nil), opts...)
+			// Warning: HACK here
+			// The base ubuntu image has this `excludes` config file which prevents
+			// installation of a lot of thigns, including doc files.
```
Comment (DannyBrito) [inline frontend/deb/distro/container.go:47]:
```suggestion
			// installation of a lot of things, including doc files.
```

File: test/azlinux_test.go [hunk 12]
```diff
@@ -1502,17 +1499,7 @@ func testCustomLinuxWorker(ctx context.Context, t *testing.T, targetCfg targetCo
...
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		// Since we also added the dep as a runtime dep, the file in the base package should be installed in the output container.
-		_, err = ref.StatFile(ctx, gwclient.StatRequest{Path: "/usr/share/doc/" + depSpec.Name + "/hello.txt"})
-		if err != nil {
-			t.Fatal(err)
-		}
+		solveT(ctx, t, gwc, sr)
```
Comment (DannyBrito) [inline test/azlinux_test.go:1502]: q: why is this deleted, the `res.SingleRef()` and err handling? Shouldn't we still do that?

Comment (cpuguy83) [reply test/azlinux_test.go:1502]: I don't think that's necessary.
If there's a problem with the build `solveT` will get it.
We don't need to check the actual content of the resulting image since this test is no longer making the assumption that the worker repos will be available at install time.

Comment (cpuguy83) [reply test/azlinux_test.go:1502]: This is mentioned in the commit where the change is at, btw.

File: frontend/deb/distro/container.go (hunk 11 above)
Comment (cpuguy83) [reply frontend/deb/distro/container.go]: Fixed.

File: test/azlinux_test.go (hunk 12 above)
Comment (DannyBrito) [reply test/azlinux_test.go:1502]: oh ok thanks, yeah, I reviewed the whole PR

**Review by DannyBrito — APPROVED** [review]
lgtm
